x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/action/TestQueryRulesetAction.java

```diff
@@ -0,0 +1,209 @@
+            builder.mapContents(matchCriteria);
+            builder.endObject();
+            builder.endObject();
+            return builder;
+        }
+
+        private static final ConstructingObjectParser<Request, String> PARSER = new ConstructingObjectParser<>(
+            "test_query_ruleset_request",
+            false,
+            (p, name) -> {
+                @SuppressWarnings("unchecked")
+                Map<String, Object> matchCriteria = (Map<String, Object>) p[0];
+                return new Request(name, matchCriteria);
+            }
+
+        );
+        static {
+            PARSER.declareObject(constructorArg(), (p, c) -> p.map(), MATCH_CRITERIA_FIELD);
+            PARSER.declareString(optionalConstructorArg(), RULESET_ID_FIELD); // Required for parsing
+        }
+
+        public static Request parse(XContentParser parser, String name) {
+            return PARSER.apply(parser, name);
+        }
+
+        @Override
+        public String[] indices() {
+            return new String[] { QueryRulesIndexService.QUERY_RULES_ALIAS_NAME };
+        }
+
+        @Override
+        public IndicesOptions indicesOptions() {
+            return IndicesOptions.lenientExpandHidden();
+        }
+
+    }
+
+    public static class Response extends ActionResponse implements ToXContentObject {
+
+        private final int totalMatchedRules;
+        private final List<MatchedRule> matchedRules;
+
+        private static final ParseField TOTAL_MATCHED_RULES_FIELD = new ParseField("total_matched_rules");
+        private static final ParseField MATCHED_RULES_FIELD = new ParseField("matched_rules");
+
+        public Response(StreamInput in) throws IOException {
+            super(in);
+            this.totalMatchedRules = in.readVInt();
+            this.matchedRules = in.readCollectionAsList(MatchedRule::new);
+        }
+
+        public Response(int totalMatchedRules, List<MatchedRule> matchedRules) {
+            this.totalMatchedRules = totalMatchedRules;
+            this.matchedRules = matchedRules;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeVInt(totalMatchedRules);
+            out.writeCollection(matchedRules, (stream, matchedRule) -> matchedRule.writeTo(stream));
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.field(TOTAL_MATCHED_RULES_FIELD.getPreferredName(), totalMatchedRules);
+            builder.startArray(MATCHED_RULES_FIELD.getPreferredName());
+            for (MatchedRule matchedRule : matchedRules) {
+                builder.startObject();
+                builder.field("ruleset_id", matchedRule.rulesetId());
+                builder.field("rule_id", matchedRule.ruleId());
+                builder.endObject();
+            }
```

Comment:
Thought: Could it be convenient for some users to add a "verbose" mode where we return the complete definition of the matched rule? We don't necessarily have to do it in this PR.

Reply:
Oh, I like that idea! I agree it should be a followup, but CC: @julianrosado that this is a possibility in the future.

Reply:
cc @mdefazio that this is a possibility in the future as well!

Reply:
ah nice! thx